### ISSUE TITLE
Relationship between mask and bbox location during training

### ISSUE THREAD

**Issue by ApoorvaSuresh** [matterport/Mask_RCNN#1500]
During the training phase, 
Do mrcnn_mask and mrcnn_bbox have details about the masks and the coordinates of the masks in the training images?

**Comment by spotiris** [comment]
It's not clear exactly what your question, but:

_mrcnn_mask_ and _mrcnn_bbox_ are the mask and bbox delta predictions for the RPN proposals which correspond to the top N=_POST_NMS_ROIS_TRAINING_  _rpn_class_ scores.
So yes, they do implicitly have details about the coordinates of the region (through the values of the RPN predictions), but no details about the mask other than the predefined _MASK_SHAPE_.

**Comment by ApoorvaSuresh** [comment]
how to derive the normalized coordinates of bbox from mrcnn_bbox? (sincemrcnn_bbox only has delta values)

**Comment by spotiris** [comment]
You will have to read the MRCNN paper for specifics, but I believe it is something along these lines:
1. Each anchor **a** has a specific (and constant) position **(y, x)** (which is normalised). The values depend on the _BACKBONE_, _BACKBONE_STRIDES_ and _RPN_ANCHOR_STRIDE_.
2. Each anchor generates lots of proposals, which as previously discussed is reduced to _POST_NMS_ROIS_TRAINING_ regions, each of which has a _proposal delta_ **delta_p**.
3. Each region is then refined in the mrcnn_bbox head again, which generated _region delta_ **delta_r**
4. Then the original coordinates for a given anchor **a* is the original coordinates, adjusted by **delta_p**, then again adjusted by **delta_r**.

Again this is for a given **anchor**, meaning you need to track down which anchor proposed the regions, which seems very complex.
I think what you are trying to do might be much more easily done by using the `unmold_detections` function, if that works for your application.

**Comment by ApoorvaSuresh** [comment]
Thank you for your reply,

yes, unmold_detections does the same. But i am trying to do this on training data.

**Comment by spotiris** [comment]
You would probably be better off modifying the network to include the positions of the anchors as an input to the keras.fit_generator function, than to try and back-calculate where the regions are located.

**Comment by ApoorvaSuresh** [comment]
How do I get the respective training bbox normalized coordinates if I include the positions of anchors into fit_generator?
